### ISSUE TITLE
Speed-Ez: Possible Issue with HR Report

### ISSUE THREAD

**Issue by kdjstudios** [kdjstudios/SABillingGitlab#1462]
In GitLab by @kdjstudios on May 1, 2019, 11:53

**Submitted by:** <joann@speedez.com>

**Helpdesk:** http://www.servicedesk.answernet.com/profiles/ticket/2019-05-01-84888/conversation

**Server:** External

**Client/Site:** Speed Ez

**Account:** 5117 The Start Center Surgical Consultants- Medical Center

**Issue:** 

I was just on the phone with Jo Ann and we were looking into their account: 5117 The Start Center Surgical Consultants- Medical Center. While doing so we were trying to confirm the imported usage vs the billable usage. For this we started to use the Human readable report; however, when trying to manual calculate the HR calculated codes we were receiving different results. I think this might be due to us having the wrong calculation formulas present in the HR report. May we please confirm that those formulas match the importing process.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 1, 2019, 11:57

@smasih @amishra @pchaudhary @tim.traylor 

I further tried to use the export option and I received the Something went wrong error. Please look into this too.

Attached I have the HR values for both the import and the calculated codes, I then tried to manual calculate them in the last column in red. As you can see they do not match and there seems to be no connection. Please look into this ASAP, as Jo Ann is wondering if we have been billing incorrectly due to this. 

Uploaded: [resource_usage_04_01_04_30.csv.20190501123157.20190501123157.xlsx](/uploads/b6cfa52872864ffd30f655bc18ba015b/resource_usage_04_01_04_30.csv.20190501123157.20190501123157.xlsx)

HR Report: [Speed_Ez_HR.xlsx](/uploads/53ba6d1c25a373997a94589685bb8047/Speed_Ez_HR.xlsx)

Marked directly for development, but please provide an estimate still. We have to have this report working regardless.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on May 10, 2019, 24:31

@kdjstudios 

Thank you for the update.


> I further tried to use the export option and I received the Something went wrong error. Please look into this too.

We have already informed you about this issue in comment #http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1388#note_38968 on  ticket #1388  


Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on May 10, 2019, 08:50

@amishra 

Thank you, please be sure to update the labels accordingly. This is a secondary ticket then to the primary now. Please post updates in the primary ticket accordingly.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on May 29, 2019, 04:51

@kdjstudios 

We have posted an estimate to fix that on my comment 
http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1388#note_39154.Please approve so that we can start the development.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 5, 2019, 10:07

Updated labels to match parent ticket

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 11, 2019, 11:22

@amishra @pchaudhary @smasih @tim.traylor 

Is there a reason why this is not released yet?

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 12, 2019, 01:23

@kdjstudios 

This is already released but didn't update the label. We request you to please check with the client to confirm.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 12, 2019, 01:58

@kdjstudios 

We would like to inform you that the main concern of the ticket is resolved along with the #1388 but there is one another concern came across while development & testing of this issue that the HR report preview Export is throwing an error. 

We have already reported that multiple times, http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1465#note_39837 ,  http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1388#note_39153.

We have also provided the estimate to fix that - http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1388#note_39154 
but you mentioned that you'll create a new ticket but we think you somehow forgot to create. 

We request you to please create a new ticket along with all of the details.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 12, 2019, 08:50

@pchaudhary @tim.traylor @smasih @amishra 

Can someone please show me how this is fixed? I just reran the HR report for the account mentioned in the ticket and the 2025 value does not appear to match the calculation provided in the HR report still. I have the orginal issue in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1465#note_39453  and also have shown the latest updates below, please let me know if I am not calculating this correctly. 

[Speed_Ez_HR_5117.xlsx](/uploads/51134bf9872bdc06c8a58ec5d821d2ce/Speed_Ez_HR_5117.xlsx)

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 13, 2019, 01:28

@kdjstudios 

We would like to inform you that the usage file you posted does not contain all the rows of the account. Please have a look at the Usage file which is uploaded on the DB -  [resource_usage_04_01_04_30.csv](/uploads/87c5ca6ca9ea8734eb02506fe8a43d67/resource_usage_04_01_04_30.csv)

* The original file contains the 4 rows of the Account.

* Please let us know if we need to do anything else.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 13, 2019, 01:29

@kdjstudios 

We have also found a ticket in which the codes changes requested by Speedez. - #746

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 13, 2019, 09:02

@pchaudhary @tim.traylor @smasih 

I am confused, are you stating that the HR report with multiple rows is not working then? The report does not make any mention of multiple rows as it should. Because the file I attached in the HR report not a usage file. 

![image](/uploads/4492f0284e383410f9b25af7c7d38e99/image.png)

Also you mention the usage file. The one http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1465#note_39453 is the one I received from the "Download" button during that month. So if we are not providing the correct download this is another bug.

In regards to the code changes there were from a year ago, while these issues are for the last few months. So that should not be relevant.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 18, 2019, 24:00

assigned to @amishra

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jun 18, 2019, 24:06

Adjusting Labels as primary

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Jun 19, 2019, 15:11

@pchaudhary @amishra 

Please read http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1465#note_42250

This is not working. Marking back for development.

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jul 5, 2019, 06:20

@kdjstudios 

We dug into the issue & below are our findings. 

> I am confused, are you stating that the HR report with multiple rows is not working then? The report does not make any mention of multiple rows as it should. Because the file I attached in the HR report, not a usage file.

* The HR report with multiple rows code is working well. 
* We checked for the Startel switch & found that the code was written in such a manner that if there are more than a row for the same resource ID's then it's getting add & then creating a single CDR record due to which it's not getting caught on the HR report but for the same scenario, the VCC switch code is creating 2 CDR records.


> Also you mention the usage file. The one http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1465#note_39453 is the one I received from the "Download" button during that month. So if we are not providing the correct download this is another bug.

* Yes, we confirm that the downloaded file for the Startel is not the same as the uploaded file.

* So, in brief, the main concern of the wrong calculation is basically not there but it's because of the wrong calculation code was updated there, however, we have fixed it now.
  * Then it seems us that the HR calculation functionality for the 2 rows is not working but it's working for all of the other switches where the CDR field is creating according to the rows but in the case of **Startel**, it's adding on its own & then creating a CDR field.
  * We have also found a bug for the downloaded file that for the **Startel**, where there were more than a row on the usage file it's getting downloaded the wrong file with the single row for the account.
there is nothing wrong with the code & the issues were only because of the calculation code was not updated.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @pchaudhary on Jul 5, 2019, 06:20

removed milestone

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Oct 29, 2019, 13:10

@amishra @smasih @jredfield 

After reviewing this today, I just want to confirm we have only two possible issues that we would like to get an update on.

1. The caluclations for the codes have been updated and are working.

2. The CDR file (SAB Download file) needs to match the uploaded file. This way there should be no reason for any discrepancies between the two files. This means if the uploaded file had 3 rows, the downloaded file also needs to show three rows and same with the HR.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 8, 2019, 04:17

@kdjstudios @jredfield 

Do we need to post an estimate before starting the development? or directly start development.

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 8, 2019, 08:37

@amishra @smasih @jredfield 

Please as always provide an estimate. Since this is a **MAJOR BUG** that needs to be fixed please automatically start with development once the estimate is posted.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 9, 2019, 04:36

@kdjstudios @jredfield 

Thank you for the update.


## Developer Estimate
|Size|Lower Time Bound|Variability|Upper Time Bound|Testing|Time for Testing| Time for Dev & Testing|
|-|-|-|-|-|-|-|
|Medium-1|4|Worked with Concepts, implementation ok|6|High Impact lots of use cases|6.0|12|

##### Estimate Notes:
* Risk areas identified: `Billing Cycle`

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 9, 2019, 11:52

@amishra @jredfield @smasih @rchauhan 

As stated in http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1465#note_50879 this has to be fixed. However, can someone please explain why it takes more time to test this then it does develop? That makes no sense.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 10, 2019, 22:22

@kdjstudios @jredfield 

Thank you for the update.

There are lots of test cases as the usage file is the most important part of SAB. A single line change may be effect more.

**I'll try to complete within the given estimate.**

Thanks

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Nov 11, 2019, 01:08

@kdjstudios @jredfield 

as you stated in your comment [here](http://gitlab.aavaz.biz/AnswerNet/SABilling/issues/1465#note_50879), it was a major bug so we fixed it immediately. 

We have fixed it and set-up a branch on the merge server for your testing. 

**Test Cases:**

1. Upload a single row for the account
   * Download Usage file should have a single row for the account
   * Match all calculations of Human Readable Report(HR Report) with the uploaded usage file and the downloaded usage file
   * Match Invoice calculations with HR Report.
2. Multiple rows for the account
   * Download Usage file should have all the uploaded rows for the account
   * Match all calculations of HR Report with the uploaded usage file and the downloaded usage file and also showing a **number of rows** uploaded 
![image](/uploads/4ed919c46bb04177a3d9a3c3ceffb72c/image.png)
   * Match Invoice calculations with HR Report.
3. Row with invalid Account   
   * Download Usage file should have status as an exception against the invalid account
4. Empty row for the Account
   * Download Usage file should have 0 as the usage for the account
   * HR Report will also show 0 against all the usages.

If there are any other test cases you can think of, please add them and test them.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 11, 2019, 10:45

@ranand @smasih @amishra @jredfield 

Thank you for the prompt update. I was able to test the upload of usage and Download of usage on the Merge server. All appears to be working correctly as they both have the same amount of rows. I then checked the HR report and it is showing the correct sum of multiple rows. I only spot checked a few accounts, but the calculation metrics appear to be correct too. As I was able to check manually for the 5195 accounts and the 2051 calculation added up

4578 = (0+0+0+1719+2859)

[resource_usage_09_01_09_30__2_.csv.20191111123105.20191111123105.xlsx](/uploads/422a8999fe044de50694c51739fe07cd/resource_usage_09_01_09_30__2_.csv.20191111123105.20191111123105.xlsx)

There however, appears to be an issue with the HR report. Either these are not taking the summed values, or something is not working. I attempted to select all of the metrics in the 2051 code and then preview and export, but it is showing zero for those on the preview and download. Also the values in the report do not appear to match the preview/export.

![image](/uploads/070a2350bca893a24b4a9c0cdfd7db42/image.png)

![image](/uploads/bb5c201cd76623288dd9257136bd1132/image.png)

[PreviewDataExportStartel.11November2019.csv](/uploads/7efa98f94842ca9fd9cf4cc61548a273/PreviewDataExportStartel.11November2019.csv)

Also it appears that we may have a mismatch between the name of metrics and their codes. For instance in the 2051 code it shows SMP:count being 7076; but on the HR it shows as "Secure Msg Resp" and I just want to confirm that is correct too.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Nov 12, 2019, 04:39

@kdjstudios 

Thank you for the Update.

Yes, you are right there is a mismatch between "Secure Msg Resp" metric and its code in the view only. For the calculations of 2051 code, "SMP:count" is being used and "Secure Msg Resp" is not being used in any usage calculations in the posting script.

As for the preview, it is printing the first row of uploaded usages. It is not summing up all the uploaded usage rows.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 12, 2019, 11:51

@ranand @amishra @smasih @jredfield 

To clarify, I see this was moved to approval, but I do not see what is needing to be approved? This need to be fixed and then put back up for testing I believe.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Nov 13, 2019, 06:09

@kdjstudios @jredfield 

Please tell us where to add SMP:count on the HR Report view? Do you want us to replace "Secure Msg Resp" with SMP:count or is there a specific place for it in the HR Report View?

And as per my understanding, 7076 code is of SMP:count and we need to fix that in the Mapping codes sheet.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 13, 2019, 14:38

@ranand 

Thank you for the clarifications. May we please add the three SPM metrics int he highlighted location below. Also as mentioned please insure the calculation are using the correct mapping codes and display accordingly in the HR too. 

![SMP_Speed_Ez](/uploads/8a82859a1b0eff51b541f83433c1e210/SMP_Speed_Ez.png)

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 13, 2019, 21:47

assigned to @ranand and unassigned @amishra

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Nov 14, 2019, 01:31

@kdjstudios @jredfield 

Thank you for the update.

We need codes for all the three SMP metrics and please clarify 7076 should be applied to which metic, i.e If 7076 is the code of "SMP:count" then what is the code for "Secure Msg Resp"?

Please provide the correct codes for **SMP:count**, **SMP:Time**, **SMP:Units** and **Secure Msg Resp**.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 14, 2019, 09:59

After looking at their Mapping file, I would expect we will need to start with the next available number. Please use the following.

[Speed-Ez_-_Human_Readable_Mapping.xlsx](/uploads/15f341dc786504af89252111126d4c6b/Speed-Ez_-_Human_Readable_Mapping.xlsx)

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Nov 15, 2019, 04:23

@kdjstudios @jredfield 

Thank you for the update.

We will let you know when we put this up for testing.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 22, 2019, 12:39

@amishra @smasih @ranand 

This is taking a bit longer in testing then I anticipated. I was hoping to have this released on this release, so that way it works properly when They do their upcoming billing cycle. However we will now probably need to do a ER release when this is ready.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 25, 2019, 06:48

@kdjstudios @jredfield 

We have set up branch on the dev2 server for your testing.

Please test it

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 25, 2019, 10:39

@amishra 

Thank you, the updated HR report looks good. I just want to confirm that this does not have the updates from #1619 where we fixed the multi line and child account importing. As I downloaded the usage file on the dev2 server and it only has one line item for the 5195 account. 

I would initially say please test this with those updates to ensure they work together and then we are approved to release this overnight.

**Comment by kdjstudios** [comment]
In GitLab by @ranand on Nov 26, 2019, 05:18

@kdjstudios 

We have tested for the multiple usage rows for the single and its working fine. 

>As I downloaded the usage file on the dev2 server and it only has one line item for the 5195 account.

It was the last CSV which we uploaded for all our testing.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 26, 2019, 15:10

@ranand @amishra @smasih 

Thank you. Please release this to production tonight.

**Comment by kdjstudios** [comment]
In GitLab by @amishra on Nov 27, 2019, 24:54

@kdjstudios @jredfield 


Thank you for the update.

We have successfully released this on the production server. Please confirm with JoAnn's.


Thanks

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Nov 27, 2019, 10:38

@amishra 

Thank you, It would appear to be working. but I posted an updated in #1619 to confirm because the SMP metrics for the last billing cycle are not showing correctly. But I presume that is because the 1#1619 will only take effect on future uploads and not past ones. Please confirm that is the case.

**Comment by kdjstudios** [comment]
In GitLab by @kdjstudios on Dec 4, 2019, 12:36

closed
